pkg/volume/awsebs/aws_ebs.go

```diff
@@ -547,21 +547,46 @@ func (c *awsElasticBlockStoreProvisioner) Provision(selectedNode *v1.Node, allow
 		pv.Spec.AccessModes = c.plugin.GetAccessModes()
 	}
 
-	requirements := make([]v1.NodeSelectorRequirement, 0)
+	deprecatedTopologyReqs := make([]v1.NodeSelectorRequirement, 0)
+	topologyReqs := make([]v1.NodeSelectorRequirement, 0)
 	if len(labels) != 0 {
 		if pv.Labels == nil {
 			pv.Labels = make(map[string]string)
 		}
 		for k, v := range labels {
 			pv.Labels[k] = v
-			requirements = append(requirements, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
+
+			if k == v1.DeprecatedLabelZoneFailureDomain || k == v1.DeprecatedLabelZoneRegion {
+				deprecatedTopologyReqs = append(deprecatedTopologyReqs, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
+				continue
+			}
+
+			if k == v1.LabelZoneFailureDomain || k == v1.LabelZoneRegion {
+				topologyReqs = append(topologyReqs, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
+				continue
+			}
+
+			// remaining labels outside known topology labels get added to both beta and GA node selectors
+			deprecatedTopologyReqs = append(deprecatedTopologyReqs, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
+			topologyReqs = append(topologyReqs, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
```

Comment:
this is pretty hard to follow... what is it trying to accomplish? it looks like a label other than zone or region will get added to both lists here, meaning both `if` blocks below evaluate to true, meaning the label gets added to NodeSelectorTerms twice. That doesn't seem right.

Reply:
same question goes for all providers

Reply:
can you describe the problem we would encounter if we left this method as-is here and didn't do label key introspection?

Reply:
Hmm, I see https://github.com/kubernetes/kubernetes/pull/81431/files#r318810458 which seems to describe this approach.

Put another way, if an external label-definer was migrating topology, how would we tell them to do it? They wouldn't be able to add special-case code here with knowledge of their label keys. I'd envision something like these steps (potentially done in distinct phases):
1. Ensure new topology labels are added to all nodes and kept in sync
2. Start using new labels exclusively in selectors
3. Resolve existing objects using old selectors
  A. Wait for them to rot and be deleted
  B. Actively update those old objects' selectors to use the new labels

This PR does 1. We could do 2 once we know the 1.17 node controller is guaranteed to be active (so probably in the 1.18 timeframe). Is 3B possible (do PV and PVCs allow updating their selector fields) or is 3A required?

Reply:
The challenge is PVs can indefinitely carry over from upgrade to upgrade. There is no guarantee they will be deleted and recreated, especially if they're used for long-running databases.  Deleting and recreating PVs and rebinding them to PVCs is a very painful process today.

For 3B, labels are allowed to be modified, but NodeAffinity is currently only [allowed to be set once](https://github.com/kubernetes/kubernetes/blob/b1ac4cda112166da7c9d0bc572fd02fc950c919a/pkg/apis/core/validation/validation.go#L1871).

Some options I see:
* Relax validation to allow NodeAffinity to be mutated, have some controller upgrade all existing PV objects once all the nodes have the new labels. If we do this, then we don't need to implement special logic in CSIMigration.
* Don't backfill existing PVs and depend on CSIMigration to do label translation in-memory. This was my current [plan](https://github.com/kubernetes/kubernetes/issues/72139). The intention was that volumes continue to provision with *only* beta labels/NodeAffinity, and the CSI migration logic will handle converting those beta labels to the provider-specific csi topology labels. However, this is going to add a dependency on CSIMigration GA for removing the beta labels (and GA of migration is slated EO2020 at the earliest).

As for when to provision with v1 labels, some options are:
* Never. Once CSIMigration is GA, then all new in-tree PVs should be provisioned with the csi driver's topology labels.
* Set v1 labels now. But to support n-2 nodes that don't have the v1 label, it needs to be added in addition to the beta labels. This is what the PR is currently doing. 
* Wait until the version where all nodes have the v1 label.  CSIMigration won't translate it, but the PV should still function fine.

Also adding @bertinatto who's working on the CSIMigration implementation. I think the decisions we make here will impact this.

Reply:
Spoke with @msau42 on slack, we agreed that the simplest/easiest solution here is to not backfill the GA labels for PVs since that support would be removed along with CSI migration anyways (because CSI plugins use their own topology labels). The downside is that we would have to support the beta labels being back-filled to nodes up until CSI migratiton is enabled by default for all in-tree volume plugins. That could be quite a few releases from now (possible after v1.21). I'm personally okay with that trade-off but would like to hear what @msau42 @davidz627 @liggitt has to say.

Reply:
it is far easier to reason about labels on nodes only, and let all things downstream of that use those labels normally with no magic/migration. New volumes could start using the new labels as soon as next release if they wanted. Keeping the beta labels as a backfill for a long while doesn't seem awful.

Reply:
Thanks Jordan, did a quick revert of the commits that change PVs, PR should be much easier to follow now -- I'll squash / clean-up commits afterwards incase we do want to add GA labels to PVs again